sogo-openchange: Support for multipart/mixed and multipart/alternative

javierag opened the pull request:
With multipart messages only one of the parts was displayed as message body.
This changeset supports both mixed and alternative multipart types.

sixstone-qq commented on OpenChange/MAPIStoreMailFolder.m:
```diff
@@ -1731,73 +1731,31 @@ - (uint32_t) exchangeRightsForRoles: (NSArray *) roles
+      message = [self lookupMessage: messageKey];
+      if (message)
+        {
+          bodyContent = [message getBodyContent];
```
This part of the code requires the body content is requested using `peek` feature [1] in order to avoid increasing the modseq on these messages and avoid setting the `\seen` flag.

See #50 and #105 for details.

Could you ensure that?

[1] https://tools.ietf.org/html/rfc3501 (Search for BODY.PEEK)

sixstone-qq commented on OpenChange/MAPIStoreMailFolder.m:
```diff
@@ -1731,73 +1731,31 @@ - (uint32_t) exchangeRightsForRoles: (NSArray *) roles
-                    }
-                }
+              [bodyData setObject: bodyContent forKey: messageKey];
             }
```
When the if contains a single statement, the brackets are not needed.

sixstone-qq commented on OpenChange/MAPIStoreMailMessage.m:
```diff
@@ -248,16 +262,74 @@ - (void) _fetchHeaderData
+      bodyPartsCharsets = [[NSMutableDictionary alloc] initWithCapacity: keysCount];
+      bodyPartsMimeTypes = [[NSMutableDictionary alloc] initWithCapacity: keysCount];
+      
+      for (i=0; i < keysCount; i++)
```
Missing spaces between `=` operator.

sixstone-qq commented on OpenChange/MAPIStoreMailMessage.m:
```diff
@@ -288,31 +360,69 @@ - (void) _fetchHeaderData
+          NSString *mimeType = [bodyPartsMimeTypes objectForKey: key];
+          content = [content bodyDataFromEncoding: encoding];
+          if ([mimeType isEqualToString: @"text/html"])
+            {
```
The brackets are not required with a single `if` statement.

javierag replied:
I know but I like to put it if the sentence to be executed spans more than one line, otherwise I get confused some times.
Should we make an exception for this cases?

sixstone-qq replied:
No pb for me

sixstone-qq commented on OpenChange/MAPIStoreMailMessage.m:
```diff
@@ -288,31 +360,69 @@ - (void) _fetchHeaderData
+                  [textContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                }
+              else
+                {
```
The same applies here.

sixstone-qq commented:
Really nice job @javierag

javierag commented:
I have modified the code to use the peek. keys in the call to [SOGoMailObjwct fetchParts:].
Also I have added some missing guards in the event of messages with missing attributes.

javierag commented:
I have fixed a event regression spotted by @blaxter . The sad thing is that I witnessed it but I forgot what was the correct display :(

I have also created the static NSArray acceptedMimetypes to avoid to create it each time a new message is parsed.

blaxter commented on OpenChange/MAPIStoreMailMessage.m:
```diff
@@ -288,31 +359,88 @@ - (void) _fetchHeaderData
+            {
+              content = [content bodyDataFromEncoding: encoding];
+              [textContent appendData: content];
+            }
```
what do you think about inserting here an else clause with an error message?

javierag replied:
i think i will do it

javierag replied:
Done

icandigitbaby commented:
:+1:

sixstone-qq commented:
Works like a charm. A major improvement here :+1:
